rfc/wallpapers_v2_network_improvements.md

```diff
@@ -0,0 +1,88 @@
+* Start date: 2022-06-21
+* RFC PR: [#112](https://github.com/mozilla-mobile/shared-docs/pull/112)
+---
+
+## Summary
+
+The wallpapers feature was introduced on Android and iOS in v98 during Q1 2022. 
+Due to late scope changes, a networking solution was quickly put together
+in allow for delivering wallpapers from a remote source. As part of a
+larger set of improvements planned for Q3 2022, it is intended to revisit
+this implementation to create a more scalable version.
+
+---
+
+## Motivation
+
+The original version of the networking solution consisted of
+adding wallpaper image files to deeply nested directories which defined the properties
+of the wallpaper. For example, an Android wallpaper might be found at:
+```
+<root>/mobile-wallpapers/android/<resolution>/<orientation>/<light/dark>/<wallpaper collection>/<wallpaper name>
+```
+
+This directory structure was confusing and difficult to manage. Copying several versions of a file to different locations
+was painful, even when automated. Automation would require consistent file naming for each new set of wallpapers, which can
+be hard to guarantee. 
+
+Additionally, downloads were started automatically during app startup if a wallpaper was missing.
+
+In general, we would like to achieve the following goals:
+
+- it is easier for product, UX, and ENG to add new wallpapers
+- it is possible to determine important metadata about a wallpaper, like its size or average color
+- download is transparent to the user
+- download is activated by the user
+- require less bandwidth, if possible
+
+---
+
+## Guide-level explanation
+
+Instead of relying on an (undocumented) convention for file paths, wallpaper metadata will be contained in a top-level JSON schema.
+Using a well-known format will help discoverability, and JSON's flexibility should allow for simpler long-term maintenance. 
+It can also provide a more direct path to any required automation. This schema could define things like:
+
+- Android file path(s)
+- iOS file path(s)
+- thumbnail path(s)
+- average color (for text and background coloring purposes)
+- collection grouping for things like seasonal promotions
+- wallpaper availability range
+- wallpaper region availability
+
+An example might look like the following:
+```json
+{
```

Comment:
Just to confirm, we are going to have an array with multiple of this wallpaper objects? 😅

Reply:
That would probably be the case if we moved forward with a definition like this, but I'm open to suggestions if you have them!

Reply:
I would personally prefer an array of multiple objects. Makes life easier.

Here's a proposal in passing: would it make sense to have the following structure:
```json
{
    "last-updated-date": "2022-01-01",
    "wallpapers": {
        // what @MatthewTighe wrote-up
    }
}
```

Having an updated date means we can very quickly check whether we need to get anything new or not. This would get updated whenever the wallpaper array gets updated. Then we don't have to parse and compare what's in the JSON with what's on disk saving some computation.

Downsides: human error. If someone forgets to update the date, but they update the wallpaper array and we skip checks based on dates, then we don't get new wallpapers until the error is caught. Doing the computation to check server JSON vs local copy regardless would ensure this never happens.

I'm literally arguing for both sides here, so I can obviously easily be swayed either way. Just throwing out a random idea and seeing if it sticks.